File: q70-q79/q74/q74.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size();
        if (row == 0) {
            return false;
        }
        int col = matrix[0].size();
        if (col == 0){
            return false;
        }
        int left = 0, right = row - 1, middle = 0;
        int row_index = 0;
        while (left <= right) {
            middle = (left + right) / 2;
            if (matrix[middle][0] == target) {
                return true;
            }
             if (matrix[middle][0] > target) {
                right = middle - 1;
            } else if (matrix[middle][0] < target) {
                left = middle + 1;
            }
        }
        if (matrix[middle][0] < target)
            row_index = middle;
        else
            row_index = middle - 1 >= 0 ? middle - 1 : 0;
        left = 0;
        right = col - 1;
        while (left <= right) {
            int middle = (left + right) / 2;
            if (matrix[row_index][middle] == target) {
                return true;
            }
            if (matrix[row_index][middle] < target) {
                left = middle + 1;
            } else if (matrix[row_index][middle] > target) {
                right = middle - 1;
            }
        }
        return false;
    }
};

File: Sequence/q1-q99/q70-q79/q74/q74.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        if len(matrix) == 0:
            return False
        if len(matrix[0]) == 0:
            return False
        left, right, middle = 0, len(matrix) - 1, 0
        while left <= right:
            middle = (left + right) // 2
            if matrix[middle][0] == target:
                return True
            if matrix[middle][0] < target:
                left = middle + 1
            elif matrix[middle][0] > target:
                right = middle - 1
        row_index = 0
        if matrix[middle][0] < target:
            row_index = middle
        else:
            if middle - 1 >= 0:
                row_index = middle - 1
            else:
                row_index = 0 
        left, right = 0, len(matrix[0]) - 1
        while left <= right:
            middle = (left + right) // 2
            if matrix[row_index][middle] == target:
                return True
            if matrix[row_index][middle] < target:
                left = middle + 1
            elif matrix[row_index][middle] > target:
                right = middle - 1
        return False
        
